ocular/src/chain/client/airdrop.rs

```diff
@@ -0,0 +1,182 @@
+                        .unwrap()
+                        .as_secs()
+                        + 60,
+                )
+                .expect("failed to derive time from system clock");
+
+                if expiration.seconds <= cutoff {
+                    return false;
+                }
+            }
+            if g.authorization.is_none() {
+                return false;
+            }
+            // I don't actually thing this is a necessary check as there is no way to specify
+            // authorization for MultiSend without using a generic one
+            let authorization = g.authorization.clone().unwrap();
+            if authorization.type_url.as_str() != GENERIC_AUTHORIZATION_URL {
+                return false;
+            }
+
+            true
+        }))
+    }
+
+    pub async fn execute_delegated_airdrop(
+        &mut self,
+        granter: AccountInfo,
+        grantee: AccountInfo,
+        fee_payer: Option<AccountId>,
+        fee_granter: Option<AccountId>,
+        payments: Vec<Payment>,
+        tx_metadata: Option<TxMetadata>,
+    ) -> Result<Response, ChainClientError> {
+        self.verify_multi_send_grant(granter.id.clone(), grantee.id.clone())
+            .await?;
+        let (inputs, outputs) =
+            multi_send_args_from_payments(granter.id.as_ref().to_string(), payments);
+        let msgs: Vec<prost_types::Any> = vec![MsgMultiSend {
+            inputs: inputs
+                .iter()
+                .map(TryInto::try_into)
+                .collect::<Result<_, _>>()?,
+            outputs: outputs
+                .iter()
+                .map(TryInto::try_into)
+                .collect::<Result<_, _>>()?,
+        }
+        .to_any()?];
+
+        self.execute_authorized_tx(grantee, fee_payer, fee_granter, msgs, tx_metadata)
+            .await
+    }
+
+    pub async fn execute_delegated_airdrop_from_toml(
+        &mut self,
+        granter: AccountInfo,
+        grantee: AccountInfo,
+        fee_payer: Option<AccountId>,
+        fee_granter: Option<AccountId>,
+        path: String,
+        tx_metadata: Option<TxMetadata>,
+    ) -> Result<Response, ChainClientError> {
+        let payments = read_payments_toml(path)?;
+        self.execute_delegated_airdrop(
+            granter,
+            grantee,
+            fee_payer,
+            fee_granter,
+            payments,
+            tx_metadata,
+        )
+        .await
+    }
+
+    pub async fn execute_airdrop(
+        &mut self,
+        sender: AccountInfo,
+        payments: Vec<Payment>,
+        tx_metadata: Option<TxMetadata>,
+    ) -> Result<Response, ChainClientError> {
+        let (inputs, outputs) =
+            multi_send_args_from_payments(sender.id.as_ref().to_string(), payments);
+        self.multi_send(sender, inputs, outputs, tx_metadata).await
+    }
+
+    pub async fn execute_airdrop_from_toml(
+        &mut self,
+        sender: AccountInfo,
```

Comment:
My assumption has been you are commenting this because you didn't realize it would all be send in one tx. If that's not the case, what's the advantage of having these account/address data fields in the toml over passing it as a param like any other transaction?

Reply:
Ah mostly my suggestion was to reduce the amount of coding needed for the airdrop and to make it easy to plug and play any combo of sender and receivers from a toml. 

E.g. The airdroper makes a script that first does a dry run, then does a small transfer to all receiver accounts, validates tx's went through, and then sends the rest of the funds. If they want to change their airdrop source address, they'd have to go in and modify the code, vs just updating their toml.

Reply:
That makes sense. Seems like a good change for this method. Similarly as for `TxMetadata`, we can probably keep fee_pay/fee_granter in it but set it with the `toml` if provided?

Reply:
I think that sounds good